internal/core/logic_test.go

```diff
@@ -260,6 +260,50 @@ func TestGetEligibleOperations(t *testing.T) {
 		"high->1200", "high->1100", // forced by MinimumFreeSize
 	)
 
+	// Critical MinimumFreeSize constraint
+	// If MinimumFreeSize is marked as critical, forced upsizes should be critical actions.
+	// Other behaviour should not be affected regardless whether the flag is set or not.
+	check(
+		"low=20%, high=80%, crit=95%, step=20%, min_free=600, min_free_is_critical=true",
+		"size=1000, usage=100",
+		"low->800", "low->700", // not restricted by critical MinimumFreeSize
+	)
+	check(
+		"low=20%, high=80%, crit=95%, step=20%, min_free=600, min_free_is_critical=false",
+		"size=1000, usage=100",
+		"low->800", "low->700", // not restricted by non-critical MinimumFreeSize
+	)
```

Comment:
The single-step variants are actually restricted by MinimumFreeSize, contrary to what the comments say. Suggestion for a non-restricted case:
```suggestion
	check(
		"low=20%, high=80%, crit=95%, step=20%, min_free=300, min_free_is_critical=true",
		"size=1000, usage=100",
		"low->800", "low->500", // not restricted by critical MinimumFreeSize
	)
	check(
		"low=20%, high=80%, crit=95%, step=20%, min_free=300, min_free_is_critical=false",
		"size=1000, usage=100",
		"low->800", "low->500", // not restricted by non-critical MinimumFreeSize
	)
```